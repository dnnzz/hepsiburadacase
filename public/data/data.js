const products = [
  {
    productId: 1,
    title: "Apple iPhone 11 Siyah",
    brand: "Apple",
    price: 120,
    discountPercentage: 10,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-04-15T01:00:00+03:00",
  },
  {
    productId: 2,
    title: "Apple iPhone 11 Yeşil",
    brand: "Apple",
    price: 1120,
    discountPercentage: 0,
    color: "Yeşil",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-11-16T01:00:00+03:00",
  },
  {
    productId: 3,
    title: "Apple iPhone 11 Beyaz",
    brand: "Apple",
    price: 300,
    discountPercentage: 20,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-07-17T01:00:00+03:00",
  },
  {
    productId: 4,
    title: "Apple iPhone 11 Kırmızı",
    brand: "Apple",
    price: 310,
    discountPercentage: 10,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/4jsB5ht/image-4.png",
    createdDate: "2021-08-18T01:00:00+03:00",
  },
  {
    productId: 5,
    title: "Apple iPhone 12 Kırmızı",
    brand: "Apple",
    price: 322,
    discountPercentage: 10,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-01-25T01:00:00+03:00",
  },
  {
    productId: 6,
    title: "Apple iPhone 12 Mavi",
    brand: "Apple",
    price: 572,
    discountPercentage: 10,
    color: "Mavi",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-02-05T05:00:00+03:00",
  },
  {
    productId: 7,
    title: "Apple iPhone 12 Beyaz",
    brand: "Apple",
    price: 521,
    discountPercentage: 10,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-03-02T08:00:00+03:00",
  },
  {
    productId: 8,
    title: "Apple iPhone 12 Mor",
    brand: "Apple",
    price: 500,
    discountPercentage: 25,
    color: "Mor",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2020-01-03T01:00:00+03:00",
  },
  {
    productId: 9,
    title: "Apple iPhone 12 Kılıf",
    brand: "Apple",
    price: 50,
    discountPercentage: 20,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-02-19T01:00:00+03:00",
  },
  {
    productId: 10,
    title: "Apple iPhone 12 Koruyucu",
    brand: "Apple",
    price: 585,
    discountPercentage: 30,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-04-05T01:00:00+03:00",
  },
  {
    productId: 11,
    title: "Samsung Note 10 Siyah",
    brand: "Samsung",
    price: 170,
    discountPercentage: 10,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-02-10T01:00:00+03:00",
  },
  {
    productId: 12,
    title: "Samsung Note 10 Yeşil",
    brand: "Samsung",
    price: 1760,
    discountPercentage: 0,
    color: "Yeşil",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-02-11T01:00:00+03:00",
  },
  {
    productId: 13,
    title: "Samsung Note 10 Beyaz",
    brand: "Samsung",
    price: 200,
    discountPercentage: 20,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-02-13T01:00:00+03:00",
  },
  {
    productId: 14,
    title: "Samsung A50 Siyah",
    brand: "Samsung",
    price: 475,
    discountPercentage: 10,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-02-14T01:00:00+03:00",
  },
  {
    productId: 15,
    title: "Samsung A10s Kırmızı",
    brand: "Samsung",
    price: 500,
    discountPercentage: 10,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/4jsB5ht/image-4.png",
    createdDate: "2021-01-01T01:00:00+03:00",
  },
  {
    productId: 16,
    title: "Samsung Note 9 Mavi",
    brand: "Samsung",
    price: 500,
    discountPercentage: 10,
    color: "Mavi",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-04-16T05:00:00+03:00",
  },
  {
    productId: 17,
    title: "Samsung Note 9 Kırmızı",
    brand: "Samsung",
    price: 500,
    discountPercentage: 10,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-06-17T08:00:00+03:00",
  },
  {
    productId: 18,
    title: "Samsung Note 9 Siyah",
    brand: "Samsung",
    price: 500,
    discountPercentage: 25,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-02-18T01:00:00+03:00",
  },
  {
    productId: 19,
    title: "Samsung Galaxy M51 Beyaz",
    brand: "Samsung",
    price: 500,
    discountPercentage: 20,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-02-19T01:00:00+03:00",
  },
  {
    productId: 20,
    title: "Samsung Note 10 Gri",
    brand: "Samsung",
    price: 500,
    discountPercentage: 30,
    color: "Gri",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-02-20T01:00:00+03:00",
  },
  {
    productId: 21,
    title: "Huawei P40 Lite Siyah",
    brand: "Huawei",
    price: 120,
    discountPercentage: 10,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-01-15T01:00:00+03:00",
  },
  {
    productId: 22,
    title: "Huawei P40 Lite Yeşil",
    brand: "Huawei",
    price: 1026,
    discountPercentage: 0,
    color: "Yeşil",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-01-16T01:00:00+03:00",
  },
  {
    productId: 23,
    title: "Huawei P40 Lite Beyaz",
    brand: "Huawei",
    price: 1220,
    discountPercentage: 11,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-01-17T01:00:00+03:00",
  },
  {
    productId: 24,
    title: "Huawei P40 Lite Kırmızı",
    brand: "Huawei",
    price: 310,
    discountPercentage: 13,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-03-02T01:00:00+03:00",
  },
  {
    productId: 25,
    title: "Huawei P Smart 2021 Kırmızı",
    brand: "Huawei",
    price: 322,
    discountPercentage: 0,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-03-04T01:00:00+03:00",
  },
  {
    productId: 26,
    title: "Huawei P Smart 2021 Mavi",
    brand: "Huawei",
    price: 572,
    discountPercentage: 15,
    color: "Mavi",
    productPhoto:"https://i.ibb.co/4jsB5ht/image-4.png",
    createdDate: "2021-03-05T05:00:00+03:00",
  },
  {
    productId: 27,
    title: "Huawei P Smart 2021 Beyaz",
    brand: "Huawei",
    price: 521,
    discountPercentage: 10,
    color: "Beyaz",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-03-06T08:00:00+03:00",
  },
  {
    productId: 28,
    title: "Huawei P Smart 2021 Mor",
    brand: "Huawei",
    price: 410,
    discountPercentage: 25,
    color: "Mor",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-03-08T01:00:00+03:00",
  },
  {
    productId: 29,
    title: "Huawei P Smart 2021 Kılıf",
    brand: "Huawei",
    price: 20,
    discountPercentage: 20,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/4jsB5ht/image-4.png",
    createdDate: "2021-03-12T01:00:00+03:00",
  },
  {
    productId: 30,
    title: "Huawei P Smart 2021 Koruyucu",
    brand: "Huawei",
    price: 285,
    discountPercentage: 30,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/Yyk7p3C/image-2.png",
    createdDate: "2021-03-15T01:00:00+03:00",
  },
  {
    productId: 31,
    title: "Nokia 301",
    brand: "Nokia",
    price: 26,
    discountPercentage: 8,
    color: "Siyah",
    productPhoto:"https://i.ibb.co/pKSBRpr/image-1.png",
    createdDate: "2021-04-10T01:00:00+03:00",
  },
  {
    productId: 32,
    title: "Xiaomi Mi 9T ",
    brand: "Xiaomi",
    price: 1285,
    discountPercentage: 30,
    color: "Kırmızı",
    productPhoto:"https://i.ibb.co/92xRQ6g/image-3.png",
    createdDate: "2021-12-12T01:00:00+03:00",
  },
];
window.localStorage.setItem("productList", JSON.stringify(products));
window.localStorage.setItem("cart", JSON.stringify([]));
