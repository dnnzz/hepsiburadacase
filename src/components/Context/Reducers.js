import { filterBySearchQuery } from "../../utils/utils";

export const cartReducer = (state, action) => {
  let updatedCart;
  switch (action.type) {
    case "ADD_TO_CART":
      updatedCart = [...state.cart, { ...action.payload, addedDate: new Date() }]
        .sort((a, b) => a.addedDate - b.addedDate)
        .reverse();
      localStorage.setItem("cart", JSON.stringify(updatedCart));
      return {
        ...state,
        cart: updatedCart,
      };
    case "REMOVE_FROM_CART":
      updatedCart = state.cart.filter((c) => c.productId !== action.payload);

      localStorage.setItem("cart", JSON.stringify(updatedCart));
      return {
        ...state,
        cart: updatedCart,
      };
    default:
      return state;
  }
};

export const productReducer = (state, action) => {
  switch (action.type) {
    case "FILTER_PRODUCTS":
      return {
        ...state,
        filteredProducts: action.payload,
      };
    case "FILTER_BY_COLOR":
      return {
        ...state,
        byColor: {
          selected: action.payload.selected,
          color: action.payload.color,
        },
      };
    case "FILTER_BY_BRAND":
      return {
        ...state,
        byBrand: {
          selected: action.payload.selected,
          brand: action.payload.brand,
        },
      };
    case "SORT_BY_PRICE_ASC":
      return {
        ...state,
        sortByPriceAsc: !state.sortByPriceAsc,
        sortByPriceDesc: false,
        sortByNewAsc: false,
        sortByNewDesc: false,
      };
    case "SORT_BY_PRICE_DESC":
      return {
        ...state,
        sortByPriceDesc: !state.sortByPriceDesc,
        sortByPriceAsc: false,
        sortByNewAsc: false,
        sortByNewDesc: false,
      };
    case "SORT_BY_NEW_ASC":
      return {
        ...state,
        sortByNewAsc: !state.sortByNewAsc,
        sortByNewDesc: false,
        sortByPriceAsc: false,
        sortByPriceDesc: false,
      };
    case "SORT_BY_NEW_DESC":
      return {
        ...state,
        sortByNewDesc: !state.sortByNewDesc,
        sortByNewAsc: false,
        sortByPriceAsc: false,
        sortByPriceDesc: false,
      };
    case "FILTER_BY_SEARCH":
      let tempSearchedProducts = [];
      tempSearchedProducts = filterBySearchQuery(state.products, action.payload);

      return { ...state, searchQuery: action.payload, filteredProducts: tempSearchedProducts };
    default:
      return state;
  }
};
