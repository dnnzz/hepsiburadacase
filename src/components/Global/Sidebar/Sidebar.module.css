.sidebarWrapper {
  text-align: left;
}

.sidebarWrapper h5 {
  font-size: 16px;
  color: #484848;
}
.sidebarWrapper ul {
  list-style: none;
  text-align: left;
  padding-inline-start: 0px;
}

.sidebarWrapper ul li {
  cursor: pointer;
  color: #646464;
  font-size: 14px;
  line-height: 24px;
}

.sidebarWrapper ul li:hover {
  color: #ff6000;
}
@media (max-width: 1900px) {
  .sidebarWrapper {
    padding-left: 50px;
    padding-right: 30px;
  }
}

@media (min-width: 1900px) {
  .sidebarWrapper {
    padding-left: 50px;
    padding-right: 25px;
  }
}
