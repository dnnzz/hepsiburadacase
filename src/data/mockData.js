export const mockAscData = [
  {
    productId: 32,
    title: "Xiaomi Mi 9T ",
    brand: "Xiaomi",
    price: 1285,
    discountPercentage: 30,
    color: "Kırmızı",
    createdDate: "2021-12-12T01:00:00+03:00",
  },
  {
    productId: 2,
    title: "Apple iPhone 11 Yeşil",
    brand: "Apple",
    price: 1120,
    discountPercentage: 0,
    color: "Yeşil",
    createdDate: "2021-11-16T01:00:00+03:00",
  },
];

export const mockDescData = [
  {
    productId: 2,
    title: "Apple iPhone 11 Yeşil",
    brand: "Apple",
    price: 1120,
    discountPercentage: 0,
    color: "Yeşil",
    createdDate: "2021-11-16T01:00:00+03:00",
  },
  {
    productId: 32,
    title: "Xiaomi Mi 9T ",
    brand: "Xiaomi",
    price: 1285,
    discountPercentage: 30,
    color: "Kırmızı",
    createdDate: "2021-12-12T01:00:00+03:00",
  },
];
