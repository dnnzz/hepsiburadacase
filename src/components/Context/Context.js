import { createContext, useContext, useReducer } from "react";
import { cartReducer, productReducer } from "./Reducers";

const AppContext = createContext();

const products = JSON.parse(localStorage.getItem("productList"));
const cart = JSON.parse(localStorage.getItem("cart"));

export const initialCartState = {
  cart: cart ? cart : [],
};
export const initialProductState = {
  products: products,
  filteredProducts: [],
  byColor: {
    selected: false,
    color: "Siyah",
  },
  byBrand: {
    selected: false,
    brand: "Apple",
  },
  sortByPriceAsc: false,
  sortByPriceDesc: false,
  sortByNewAsc: false,
  sortByNewDesc: false,
  searchQuery: "",
};
export const Context = ({
  children,
  testState,
  testProductState,
  testproductDispatch,
  testDispatch,
}) => {
  // get Localstorage productlist ------

  const [iState, iDispatch] = useReducer(cartReducer, initialCartState);

  const [iProductState, iProductDispatch] = useReducer(productReducer, initialProductState);

  const state = testState ? testState : iState;
  const dispatch = testDispatch ? testDispatch : iDispatch;
  const productState = testProductState ? testProductState : iProductState;
  const productDispatch = testproductDispatch ? testproductDispatch : iProductDispatch;

  return (
    <AppContext.Provider value={{ state, dispatch, productState, productDispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export const AppState = () => {
  return useContext(AppContext);
};

export default Context;
