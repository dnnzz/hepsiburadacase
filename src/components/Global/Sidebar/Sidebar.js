/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from "react";
import styles from "./Sidebar.module.css";
import { AppState } from "../../Context/Context";
import { getCountByProperty, filterProducts } from "../../../utils/utils";
export default function Sidebar() {
  const {
    productState: { products, filteredProducts, byColor, byBrand },
    productDispatch,
  } = AppState();

  const [colorQty, setColorQty] = useState({});
  const [brandQty, setBrandQty] = useState({});

  const handleClickSortElements = (type) => {
    productDispatch({
      type: type,
    });
  };

  /* 
  check for active filters if we got filter then refilter from existing filteredProducts
  */
  useEffect(() => {
    let payloadVal;
    if (filteredProducts.length > 0) {
      payloadVal = filterProducts(products, byColor, byBrand);
      productDispatch({
        type: "FILTER_PRODUCTS",
        payload: payloadVal,
      });
    } else {
      payloadVal = filterProducts(products, byColor, byBrand);
      productDispatch({
        type: "FILTER_PRODUCTS",
        payload: payloadVal,
      });
    }
  }, [byBrand.selected, byColor.selected]);

  /* 
  Every time productList array changes useEffect will evaluate
  getCountByProperty() function feeds filter section
  and check for active filter if we got active filter counts calculated from filteredProducts 
  */
  useEffect(() => {
    if (filteredProducts.length > 0) {
      setColorQty(getCountByProperty(filteredProducts, "color"));
      setBrandQty(getCountByProperty(filteredProducts, "brand"));
    } else {
      setColorQty(getCountByProperty(products, "color"));
      setBrandQty(getCountByProperty(products, "brand"));
    }
  }, [products, filteredProducts]);

  return (
    <div className={styles.sidebarWrapper}>
      <h5>Renk</h5>
      <ul>
        {/* Takes colorQty state and prints key&values to filter section */}
        {/* if User click again on same filter key - filter will be removed at reducer  
            Check happens here -> byColor.selected && byColor.color === key  
             then send payload false and revaluate productList
             if -> byColor.selected false - then send payload true and filter productList 
            */}
        {Object.entries(colorQty).map(([key, value]) => {
          return (
            <li
              onClick={(e) => {
                byColor.selected && byColor.color === key
                  ? productDispatch({
                      type: "FILTER_BY_COLOR",
                      payload: {
                        color: key,
                        selected: false,
                      },
                    })
                  : productDispatch({
                      type: "FILTER_BY_COLOR",
                      payload: {
                        color: key,
                        selected: true,
                      },
                    });
              }}
              key={key}
            >
              {key} ({value})
            </li>
          );
        })}
      </ul>
      <h5>Sıralama</h5>
      <ul>
        <li
          onClick={(e) => {
            handleClickSortElements("SORT_BY_PRICE_ASC");
          }}
        >
          En düşük fiyat
        </li>
        <li
          onClick={(e) => {
            handleClickSortElements("SORT_BY_PRICE_DESC");
          }}
        >
          En yüksek fiyat
        </li>
        <li
          onClick={(e) => {
            handleClickSortElements("SORT_BY_NEW_ASC");
          }}
        >
          En yeniler (A&gt;Z)
        </li>
        <li
          onClick={(e) => {
            handleClickSortElements("SORT_BY_NEW_DESC");
          }}
        >
          En yeniler (Z&gt;A)
        </li>
      </ul>
      <h5>Marka</h5>
      <ul>
        {/* Takes brandQty state and prints key&values to filter section */}
        {/* if User click again on same filter key - filter will be removed at reducer 
             if -> byBrand.selected true - then send payload false 
             if -> byBrand.selected false - then send payload true and filter productList 
            */}
        {Object.entries(brandQty).map(([key, value]) => {
          return (
            <li
              onClick={(e) => {
                byBrand.selected && byBrand.brand === key
                  ? productDispatch({
                      type: "FILTER_BY_BRAND",
                      payload: {
                        brand: key,
                        selected: false,
                      },
                    })
                  : productDispatch({
                      type: "FILTER_BY_BRAND",
                      payload: {
                        brand: key,
                        selected: true,
                      },
                    });
              }}
              key={key}
            >
              {key} ({value})
            </li>
          );
        })}
      </ul>
    </div>
  );
}
