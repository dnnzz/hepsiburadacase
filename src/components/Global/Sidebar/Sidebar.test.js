import React from "react";

import { render, screen} from "@testing-library/react";
import { Context } from "../../Context/Context";
import "@testing-library/jest-dom";
import Sidebar from "./Sidebar";
import {products} from '../../../data/data';
import userEvent from "@testing-library/user-event";

const testState = {
  products: products,
  filteredProducts: [],
  byColor: {
    selected: false,
    color: "Siyah",
  },
  byBrand: {
    selected: false,
    brand: "Apple",
  },
  sortByPriceAsc: false,
  sortByPriceDesc: false,
  sortByNewAsc: false,
  sortByNewDesc: false,
  searchQuery: "",
};
const dispatch = jest.fn();
const productDispatch = jest.fn();
function renderSidebar() {
  
  return render(
    <Context
    testDispatch={dispatch}
    testState={{cart:[]}}
    testproductDispatch={productDispatch}
    testProductState={testState}
  >
    <Sidebar />
  </Context>,
  );
}

describe("sidebar tests", () => {
  beforeEach(()=> renderSidebar())
  it("renders sidebar to screen", () => {
    expect(screen.getByTestId("sidebar")).toBeInTheDocument();
  });
  it("clicks filterbycolor and dispatch action",()=>{
    const filterButton = screen.getByTestId("Siyah");
    userEvent.click(filterButton);
    expect(productDispatch).toHaveBeenCalledWith({ type: "FILTER_BY_COLOR" , payload: {
      color: "Siyah",
      selected: true,
    }});
  })
  it("clicks filterbybrand and dispatch action",()=>{
    const filterButton = screen.getByTestId("Apple");
    userEvent.click(filterButton);
    expect(productDispatch).toHaveBeenCalledWith({ type: "FILTER_BY_BRAND" , payload: {
      brand: "Apple",
      selected: true,
    }});
  })
  it("clicks sortelement and dispatch action",()=>{
    const sortButton = screen.getByTestId("sortByPriceAsc");
    userEvent.click(sortButton);
    expect(productDispatch).toHaveBeenCalledWith({type:"SORT_BY_PRICE_ASC"})
  })
});
