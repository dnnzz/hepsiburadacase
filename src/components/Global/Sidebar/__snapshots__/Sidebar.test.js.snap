// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render sidebar renders sidebar to screen 1`] = `ShallowWrapper {}`;
